File: src/video/services/asset_processing_specs.py
"""Asset processing specifications for lineup-ready media.

Defines the target specifications for each asset type when processing
from "raw" (AI-generated / uploaded) to "lineup-ready" (standardized).

Asset Types:
  fullbody    — Full body in tenue (PNG, 1080×1920, transparent bg)
  closeup     — Close-up portrait (PNG, 512×512, transparent bg)
  intro       — Short intro video (WebM VP9, 540×960, 25fps, transparent bg via RVM)
  celebration — Celebration video (WebM VP9, 540×960, 25fps, transparent bg via RVM)
  then_vs_now — Then-vs-now video (WebM VP9, 540×960, 25fps, transparent bg via RVM)

Variant Value Structure:
  Old: images.fullbody.home = "s3://path/to/image.png"
  New: images.fullbody.home = {
      "raw": "s3://...",
      "processed": "s3://...",
      "processing_state": "processed",
      "specs": {"width": 1080, "height": 1920, "format": "png", "bg_removed": true}
  }
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any


class ProcessingState(str, Enum):
    """Asset processing states."""

    RAW = "raw"
    PROCESSING = "processing"
    CANCELLING = "cancelling"
    CANCELLED = "cancelled"
    PROCESSED = "processed"
    FAILED = "failed"


@dataclass(frozen=True)
class ImageSpec:
    """Target spec for image assets."""

    width: int
    height: int
    format: str = "png"
    bg_removed: bool = True
    aspect_ratio: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "type": "image",
            "width": self.width,
            "height": self.height,
            "format": self.format,
            "bg_removed": self.bg_removed,
            "aspect_ratio": self.aspect_ratio,
        }


@dataclass(frozen=True)
class VideoSpec:
    """Target spec for video assets."""

    width: int
    height: int
    format: str = "mp4"
    fps: int = 30
    codec: str = "h264"
    bg_removed: bool = True
    max_duration: float | None = None  # None = keep original
    aspect_ratio: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "type": "video",
            "width": self.width,
            "height": self.height,
            "format": self.format,
            "fps": self.fps,
            "codec": self.codec,
            "bg_removed": self.bg_removed,
            "max_duration": self.max_duration,
            "aspect_ratio": self.aspect_ratio,
        }


# ── Spec Definitions ─────────────────────────────────────────────────────────

FULLBODY_SPEC = ImageSpec(
    width=1080,
    height=1920,
    format="png",
    bg_removed=True,
    aspect_ratio="9:16",
)

CLOSEUP_SPEC = ImageSpec(
    width=512,
    height=512,
    format="png",
    bg_removed=True,
    aspect_ratio="1:1",
)

INTRO_SPEC = VideoSpec(
    width=540,
    height=960,
    format="webm",
    fps=25,
    codec="vp9",
    bg_removed=True,
    max_duration=None,
    aspect_ratio="9:16",
)

CELEBRATION_SPEC = VideoSpec(
    width=540,
    height=960,
    format="webm",
    fps=25,
    codec="vp9",
    bg_removed=True,
    max_duration=None,
    aspect_ratio="9:16",
)

THEN_VS_NOW_SPEC = VideoSpec(
    width=540,
    height=960,
    format="webm",
    fps=25,
    codec="vp9",
    bg_removed=True,
    max_duration=None,
    aspect_ratio="9:16",
)

PHOTO_COMPOSITE_SPEC = VideoSpec(
    width=540,
    height=960,
    format="webm",
    fps=25,
    codec="vp9",
    bg_removed=True,
    max_duration=None,
    aspect_ratio="9:16",
)

WALKING_COMPOSITE_SPEC = VideoSpec(
    width=540,
    height=960,
    format="webm",
    fps=25,
    codec="vp9",
    bg_removed=True,
    max_duration=None,
    aspect_ratio="9:16",
)

ASSET_SPECS: dict[str, ImageSpec | VideoSpec] = {
    "fullbody": FULLBODY_SPEC,
    "closeup": CLOSEUP_SPEC,
    "intro": INTRO_SPEC,
    "celebration": CELEBRATION_SPEC,
    "then_vs_now": THEN_VS_NOW_SPEC,
    "photo_composite": PHOTO_COMPOSITE_SPEC,
    "walking_composite": WALKING_COMPOSITE_SPEC,
}


# ── Variant Value Helpers ────────────────────────────────────────────────────


def normalize_variant_value(value: str | dict | None) -> dict | None:
    """Normalize a variant value to the new object format.

    Handles both old format (plain URL string) and new format (dict with
    raw/processed keys).

    Returns:
        Normalized dict or None if empty.
    """
    if not value:
        return None

    if isinstance(value, str):
        return {
            "raw": value,
            "processed": None,
            "processing_state": ProcessingState.RAW.value,
        }

    if isinstance(value, dict):
        # Already new format — ensure required keys
        return {
            "raw": value.get("raw", ""),
            "processed": value.get("processed"),
            "processed_source": value.get("processed_source"),  # full-quality alpha file
            "processing_state": value.get("processing_state", ProcessingState.RAW.value),
            "specs": value.get("specs"),
            "error": value.get("error"),
            "processed_at": value.get("processed_at"),
        }

    return None


def get_best_url(value: str | dict | None) -> str | None:
    """Get the best available URL (prefers processed, falls back to raw).

    NOTE: For browser playback. Returns preview MP4 when available.
    For FFmpeg/lineup composition, use get_ffmpeg_best_url() instead.
    """
    normalized = normalize_variant_value(value)
    if not normalized:
        return None
    return normalized.get("processed") or normalized.get("raw") or None


def get_ffmpeg_best_url(value: str | dict | None) -> str | None:
    """Get the best URL for FFmpeg/lineup composition (prefers alpha source).

    Priority order:
    1. processed_source - full-quality file with alpha (e.g. ProRes .mov)
    2. processed - browser-playable version (may be preview MP4 without alpha)
    3. raw - original uploaded file

    Use this for FFmpeg pipelines that need the alpha channel.
    """
    normalized = normalize_variant_value(value)
    if not normalized:
        return None
    return (
        normalized.get("processed_source")
        or normalized.get("processed")
        or normalized.get("raw")
        or None
    )


def get_lineup_ready_url(value: str | dict | None) -> str | None:
    """Get the lineup-ready URL (processed only, no fallback)."""
    normalized = normalize_variant_value(value)
    if not normalized:
        return None
    return normalized.get("processed") or None


def is_lineup_ready(value: str | dict | None) -> bool:
    """Check if an asset variant has a processed (lineup-ready) version."""
    normalized = normalize_variant_value(value)
    if not normalized:
        return False
    return normalized.get("processing_state") == ProcessingState.PROCESSED.value and bool(
        normalized.get("processed")
    )

File: demo/src/constants/assetProcessingSpecs.ts
/**
 * Asset Processing Specifications
 *
 * Defines the target specs for each asset type when processing from "raw" to "lineup-ready".
 * Used by both the frontend (to display specs & badges) and referenced by the backend processor.
 */

// ── Processing States ────────────────────────────────────────────────────────

export type AssetProcessingState = 'raw' | 'processing' | 'cancelling' | 'cancelled' | 'processed' | 'failed';

// ── Asset Spec Definitions ───────────────────────────────────────────────────

export interface ImageProcessingSpec {
  type: 'image';
  /** Target width in pixels */
  width: number;
  /** Target height in pixels */
  height: number;
  /** Output format (always PNG for transparency support) */
  format: 'png';
  /** Whether background must be removed */
  bgRemoved: true;
  /** Aspect ratio label for display */
  aspectRatio: string;
}

export interface VideoProcessingSpec {
  type: 'video';
  /** Target width in pixels */
  width: number;
  /** Target height in pixels */
  height: number;
  /** Output format */
  format: 'mp4' | 'webm';
  /** Frames per second */
  fps: number;
  /** Whether background must be removed (chroma-key or AI) */
  bgRemoved: true;
  /** Max duration in seconds (null = keep original) */
  maxDuration: number | null;
  /** Video codec */
  codec: 'h264' | 'vp9';
  /** Aspect ratio label for display */
  aspectRatio: string;
}

export type ProcessingSpec = ImageProcessingSpec | VideoProcessingSpec;

// ── Target Specifications per Asset Type ─────────────────────────────────────

/**
 * Full body (In Tenue) — PNG, 9:16, transparent background
 * Used as: reveal scene, hold scene in lineup video
 */
export const FULLBODY_SPEC: ImageProcessingSpec = {
  type: 'image',
  width: 1080,
  height: 1920,
  format: 'png',
  bgRemoved: true,
  aspectRatio: '9:16',
};

/**
 * Close-up — PNG, 1:1 square, transparent background
 * Used as: positioned overlay in lineup video
 */
export const CLOSEUP_SPEC: ImageProcessingSpec = {
  type: 'image',
  width: 512,
  height: 512,
  format: 'png',
  bgRemoved: true,
  aspectRatio: '1:1',
};

/**
 * Short Intro video — WebM VP9, 9:16, transparent background via RVM
 * Used as: player intro segment in lineup video
 */
export const INTRO_SPEC: VideoProcessingSpec = {
  type: 'video',
  width: 1080,
  height: 1920,
  format: 'webm',
  fps: 30,
  bgRemoved: true,
  maxDuration: null, // keep original duration
  codec: 'vp9',
  aspectRatio: '9:16',
};

/**
 * Celebration video — WebM VP9, 9:16, transparent background via RVM
 * Used as: goal celebration segment
 */
export const CELEBRATION_SPEC: VideoProcessingSpec = {
  type: 'video',
  width: 1080,
  height: 1920,
  format: 'webm',
  fps: 30,
  bgRemoved: true,
  maxDuration: null,
  codec: 'vp9',
  aspectRatio: '9:16',
};

// ── Lookup Map ───────────────────────────────────────────────────────────────

export const ASSET_PROCESSING_SPECS: Record<string, ProcessingSpec> = {
  fullbody: FULLBODY_SPEC,
  closeup: CLOSEUP_SPEC,
  intro: INTRO_SPEC,
  celebration: CELEBRATION_SPEC,
} as const;

// ── Variant Value Interface ──────────────────────────────────────────────────

/**
 * New structured variant value — replaces plain URL string.
 *
 * Migration path:
 *   Old: images.fullbody.home = "s3://path/to/image.png"
 *   New: images.fullbody.home = { raw: "s3://...", processed: "s3://...", ... }
 *
 * Backward compat: if the value is a string, treat it as { raw: string }.
 */
export interface AssetVariantValue {
  /** Original AI-generated or uploaded file */
  raw: string;
  /** Lineup-ready processed file (null if not yet processed) */
  processed: string | null;
  /** Current processing state */
  processing_state: AssetProcessingState;
  /** Browser-playable preview URL (e.g. MP4 when processed is MOV) */
  preview_url?: string;
  /** Original processed file before browser conversion (e.g. ProRes MOV) */
  processed_source?: string;
  /** Actual specs of the processed file (set after processing) */
  specs?: {
    width?: number;
    height?: number;
    format?: string;
    bg_removed?: boolean;
    duration?: number;
    fps?: number;
  };
  /** Error message if processing failed */
  error?: string;
  /** ISO timestamp of last processing attempt */
  processed_at?: string;
}

// ── Helper Functions ─────────────────────────────────────────────────────────

/**
 * Normalize a variant value from either old string format or new object format.
 */
export function normalizeVariantValue(value: string | AssetVariantValue | null | undefined): AssetVariantValue | null {
  if (!value) return null;

  // Old format: plain URL string → treat as raw
  if (typeof value === 'string') {
    return {
      raw: value,
      processed: null,
      processing_state: 'raw',
    };
  }

  // New format: already an object
  return value;
}

/**
 * Get the best available URL for display (prefers processed, falls back to raw).
 */
export function getBestUrl(value: string | AssetVariantValue | null | undefined): string | null {
  const normalized = normalizeVariantValue(value);
  if (!normalized) return null;
  // Prefer browser-playable preview (e.g. MP4) over processed (may be MOV)
  return normalized.preview_url || normalized.processed || normalized.raw || null;
}

/**
 * Get the lineup-ready URL (processed only, no fallback).
 */
export function getLineupReadyUrl(value: string | AssetVariantValue | null | undefined): string | null {
  const normalized = normalizeVariantValue(value);
  if (!normalized) return null;
  return normalized.processed || null;
}

/**
 * Check if an asset variant is lineup-ready.
 */
export function isLineupReady(value: string | AssetVariantValue | null | undefined): boolean {
  const normalized = normalizeVariantValue(value);
  if (!normalized) return false;
  return normalized.processing_state === 'processed' && !!normalized.processed;
}

/**
 * Check if an asset variant is currently being processed.
 */
export function isProcessing(value: string | AssetVariantValue | null | undefined): boolean {
  const normalized = normalizeVariantValue(value);
  if (!normalized) return false;
  return normalized.processing_state === 'processing' || normalized.processing_state === 'cancelling';
}

/**
 * Human-readable label for processing state.
 */
export function getProcessingStateLabel(state: AssetProcessingState | null | undefined): { label: string; color: string; icon: string } {
  switch (state) {
    case 'raw':
      return { label: 'Ruw', color: '#f59e0b', icon: '🔶' };
    case 'processing':
      return { label: 'Bezig...', color: '#3b82f6', icon: '⏳' };
    case 'cancelling':
      return { label: 'Annuleren...', color: '#3b82f6', icon: '⏹️' };
    case 'cancelled':
      return { label: 'Geannuleerd', color: '#f59e0b', icon: '⏹️' };
    case 'processed':
      return { label: 'Lineup-ready', color: '#10b981', icon: '✅' };
    case 'failed':
      return { label: 'Fout', color: '#ef4444', icon: '❌' };
    default:
      // null/undefined means no processing state yet - treat as raw
      return { label: 'Ruw', color: '#f59e0b', icon: '🔶' };
  }
}
